proxy/src/metrics.rs

```diff
@@ -30,7 +32,7 @@ pub async fn task_main(config: &MetricCollectionConfig) -> anyhow::Result<()> {
         info!("metrics collector has shut down");
     }
 
-    let http_client = http::new_client();
```

Comment:
I see a potential issue in consumption reporting logic, which can become worse with timeouts.

In `collect_metrics_iteration` the logic is the following:
1. take `proxy_io_bytes_per_client` value accumulated in-memory since the last restart
2. get a difference between `cached_metrics` and `proxy_io_bytes_per_client` value and send it to the `metric_collection_endpoint`
3. update `cached_metrics` if the response is successful

The issue is that 3. updates the reported value only if we got a **successful** response. In theory it's possible that HTTP request was received and processed by the `metric_collection_endpoint`, but response to the proxy didn't arrive on time (or proxy got an error response).

In this case, proxy will try to report these traffic bytes again on the next iteration, and user may be charged 2+ times for the same traffic.

This issue was already here, because network connections always can break in unexpected places and client won't get a response. But with timeouts this issue is probably more likely, because it can be easily reproduced when `metric_collection_endpoint` is overloaded and very slow.


Please note that I may have missed something, I haven't looked at consumption code before today, so maybe we already had this discussion somewhere else.

Reply:
@petuhovskiy  IIRC that's why we have ```idempotency_key``` to handle such cases, will let @lubennikovaav to keep me honest here

Reply:
AFAIU `idempotency_key` is useful and helps in case of retries, but doesn't work in "overloaded" case (console processed the consumption but response didn't get back multiple times).

Note that this issue is surely not common yet, I've seen only ~30 errors in proxy consumption logs for the last month, and all of them look like this:
```
2023-06-26T13:07:55.679756Z ERROR collect_metrics_iteration: metrics endpoint refused the sent metrics: Response { url: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("neon-internal-api.aws.neon.tech")), port: None, path: "/billing/api/v1/usage_events", query: None, fragment: None }, status: 500, headers: {"date": "Mon, 26 Jun 2023 13:07:55 GMT", "content-type": "application/json", "content-length": "41", "connection": "keep-alive"} }
```

Reply:
uh(
indeed there is a problem with "update cached_metrics if the response is successful" logic.

We can change it to "update cached metrics always" and rely on retries to deliver metrics successfully.
Then in the worst case we'll under bill for undelivered deltas.

Reply:
done, although I would rather avoid combining such changes in the same PR